### PR TITLE
std.Uri: Assume query string is already escaped

### PR DESCRIPTION
**Pull request by dadrian**
std.Uri.format currently calls writeEscapedQuery when writing out the query string. Unfortunately, this means that it is not possible to specify a query that contains an ampersand or an equals sign, since these are left unescaped by the writeEscapedQuery.

For example, assume you are attempting to pass the value `M&M` as a query variable named `candy`, and the value `=` as the variable `cmp`. Note that the escaped value for `=` is `%3D`, `&` is `%26`, and `%` is `%25`.

Consider the possible values for Uri.query attempting to encode the above example:
1. Unescaped: `candy=M&M&cmp==`
2. Escaped: `candy=M%26M&cmp=%3D`

After passing through writeEscapedQuery, (1) encodes to itself, because `=` and `&` are valid characters in a query string. However, the semantic meaning of that string unclear, but at minimum contains the set `{ candy: "M", M: "" }`, and the behavior around `cmp` is dependent on any rules that may or may not exist for interpreting malformed query values.

On the other hand, (2) would end up double-escaped because '%' gets escaped to '%25', resulting in `candy=M%2526M&cmp=%253D`, which is the set `{ candy: "M%26M", cmp: "%3D"}`, which is also wrong.

Ultimately, if there's any instances of `=` or `&` in the query variable, not only is writeEscapedQuery incorrect, but there's no way to write a correct function that operates on a combined query string, rather than a set of key/value pairs.

Instead, Uri should assume the query string is already escaped. Callers can use Uri.escapeString and Uri.writeEscapedString to encode their query values individually, and then combine.

This approach is similar to the Go standard library, which uses a URL.RawQuery field to store the encoded query as a string.

### CODE REVIEW

**Comment by dadrian** [comment]
It might be useful to have a function that takes a list of K/V pairs, or a map, and encodes them as a query string, similar to https://pkg.go.dev/net/url#Values. However, I'm not sure what the standard patterns, if any, are in the stdlib for tuples or maps, as they all seem fairly allocation-heavy.

**Comment by nektro** [comment]
I have https://github.com/nektro/zig-UrlValues, welcome to inspect and take

**Comment by MasterQ32** [comment]
If unchanged from my impl, this is intentional.

There is explicit functions to escape query components, same as other url apis do, as query format can be arbitrary, but still needs escaping for certain characters.

if values need to be escaped, this has to be done explicitly beforehand

**Review by andrewrk — CHANGES_REQUESTED** [review]
@MasterQ32 you seem to be saying the same thing as @dadrian.

Looking at the commit log, I see a bunch of suspicious query string escaping related things. I think perhaps some well meaning contributors have taken the code in a problematic direction.

@dadrian would you be willing to rebase this on master, and then rework the query string escaping to avoid these pitfalls, perhaps enhancing the comments to make the code more resistant to well-meaning but ultimately destructive contributions in the future?

File: lib/std/Uri.zig [hunk 1]
```diff
@@ -251,7 +243,7 @@ pub fn format(
 
         if (uri.query) |q| {
```
Comment (MasterQ32) [inline lib/std/Uri.zig:244]: If we do this change, i'd rename `query` to `raw_query` to inform the user that it is actually not containing any escaped data

**Review by MasterQ32 — CHANGES_REQUESTED** [review]

**Comment by MasterQ32** [comment]
Another thing i'd maybe change is making a validate function people can assert on that will check if all URI components will behave in a good manner (so `raw_uri` doesn't contain bad characters and such)

**Comment by dadrian** [comment]
Apologies for the delay. I'm happy to get this rebased.

**Comment by andrewrk** [comment]
Closing since this is untouched in 3 months and is not passing CI checks. Please make a bug report first to continue this effort so that we can agree on the expected API and desired unit test behavior.

**Pull request closed without merging**